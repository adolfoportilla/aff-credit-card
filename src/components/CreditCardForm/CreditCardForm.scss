@import '../../scss/base.scss';

::placeholder {
  // color: grey;
  // color: #7f8c8d;
  // color: #bdc3c7;
  // color: #68737a; darker
  color: #848f96;
}

.credit-card-form {
  background-color: $affirm-dark-pink;
  // background-color: $background;
  border-radius: 5px;
  padding-top: 50px;
  padding-bottom: 50px;
  box-shadow: 0 7px 14px rgba(50, 50, 93, 0.1), 0 3px 6px rgba(0, 0, 0, 0.08);

  .credit-card-input {
    height: 35px;
    border-radius: 8px;
    // background: #34495e;
    background: $affirm-pink;
    border-style: solid;
    border-color: $affirm-pink;
    // color: #4e565b;
    color: blue;

    &.half {
      width: calc(50% - 5px);
    }
  }
  .credit-card-button {
    width: 150px;
    height: 30px;
    border-radius: 5px;
    background: $affirm;
    color: $white;
      &:active {
        // border: 1px solid $primary;
        // background-color: $affirm-green-secondary;
        // background-color: $affirm-lighter;
        background-color: #0d4e9c;
        // background-color: #62a3f1;
      }
  }

  .error {
    @extend  .m-t-med, .m-l-xxl, .m-r-xxl, .p-t-sm, .p-b-sm;
    // @extend  .m-t-med,  .p-t-sm, .p-b-sm;
    color: $white;
    // font-size: 0.85rem;
    margin-bottom: -36.5px;
    background: $background;
    // background: #f5dadc;
    border-radius: 5px;
    // outline: ;

    .error-message {
      @extend .m-t-none, .m-b-none;
      color: #E95757;
      // color: white;
      font-family: Arial, Helvetica, sans-serif;
      font-weight: 500;
      font-size: 14px;
      // margin-left: 2px;

    }

  }
}
