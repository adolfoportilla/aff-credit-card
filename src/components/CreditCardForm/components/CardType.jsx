import React from "react";

import { ReactComponent as Visa } from "payment-icons/min/flat/visa.svg";
import { ReactComponent as Amex } from "payment-icons/min/flat/amex.svg";
import { ReactComponent as Default } from "payment-icons/min/mono/default.svg";

import "./CardType.scss";

const CardType = ({ type, wrapperClassName }) => {
  switch (type) {
    case "AMEX":
      return (
        <div className={wrapperClassName}>
          <Amex className="card-type" />
        </div>
      );
    case "VISA":
      return (
        <div className={wrapperClassName}>
          <Visa className="card-type" />
        </div>
      );
    default:
      return (
        <div className={wrapperClassName}>
          <Default className="card-type" />
        </div>
      );
  }
};

export default CardType;
