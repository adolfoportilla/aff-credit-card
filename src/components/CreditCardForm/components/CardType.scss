@import '../../../scss/base.scss';

.card-type {
  margin-top: 7.5px;
  margin-left: 7.5px;
  height: 20px;
  width: auto;
}
