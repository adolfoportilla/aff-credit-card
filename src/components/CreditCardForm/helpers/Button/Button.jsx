import React from "react";

import "./Button.scss";

const Button = ({ wrapperClassName, baseClassName, text, onClick }) => {
  let buttonClassName = "button";
  if (baseClassName) {
    buttonClassName += ` ${baseClassName}`;
  }
  return (
    <div className={wrapperClassName}>
      <button onClick={onClick} className={buttonClassName}>
        {text}
      </button>
    </div>
  );
};

export default Button;
