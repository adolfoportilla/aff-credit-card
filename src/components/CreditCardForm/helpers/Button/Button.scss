@import '../../../../scss/base.scss';

.button {
  &:hover {
    background-color: #3588ee;
  }
}

button {
  cursor: pointer;
  border: none;
  background: none;
  transition: color 0.25s ease, fill 0.25s ease;
  &:hover, &:active, &:focus {
    background-color: blue;
    text-decoration: none;
  }
}
