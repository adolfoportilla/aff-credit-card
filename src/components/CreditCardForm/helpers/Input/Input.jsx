import React from "react";

import "./Input.scss";

const Input = ({
  autocomplete,
  baseClassName,
  inputMode,
  maxLength,
  name,
  onChange,
  placeholder,
  type,
  value,
  wrapperClassName
}) => {
  return (
    <div className={wrapperClassName}>
      <input
        type={type}
        name={name}
        onChange={onChange}
        placeholder={placeholder}
        className={baseClassName}
        maxLength={maxLength}
        value={value}
        inputMode={inputMode}
        autoComplete={autocomplete}
      />
    </div>
  );
};

export default Input;
