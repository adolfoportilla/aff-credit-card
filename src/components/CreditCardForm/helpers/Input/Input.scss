@import '../../../../scss/base.scss';

input {
  border: none;
  background: transparent !important;
  padding-left: 6px;
  width: 100%;

}

::placeholder {
  color: #bdc3c7;
}
