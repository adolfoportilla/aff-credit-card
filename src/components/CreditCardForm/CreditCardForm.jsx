import React from "react";
import { IoIosCloseCircle } from "react-icons/io";
import { FiAlertTriangle as Icon } from "react-icons/fi";

import Button from "./helpers/Button/Button";
import CardType from "./components/CardType";
import creditCards from "./cards";
import Input from "./helpers/Input/Input";

import "./CreditCardForm.scss";

const staticText = {
  button: "PAY",
  namePlaceholder: "Name",
  numberPlaceholder: "1234 1234 1234 1234",
  expiryPlaceholder: "MM / YY",
  CVCPlaceholder: "CVC"
};

class CreditCardForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      cardType: creditCards.types.DEFAULT,
      // errorMessage: "remove this before submitting"
      errorMessage: ""
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.isValidForm = this.isValidForm.bind(this);
    this.clearForm = this.clearForm.bind(this);
    this.renderInputs = this.renderInputs.bind(this);
    this.displayErrorMessage = this.displayErrorMessage.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(e) {
    const [targetValue, targetName] = [e.target.value, e.target.name];

    const cardValidator = Object.values(creditCards.validators).find(card =>
      card.isCardType(targetValue)
    );

    console.log("------------------");
    console.log("called");
    console.log("------------------");

    switch (targetName) {
      case "number":
        const newCardType = cardValidator.type;
        const formattedNumber = cardValidator.formatNumber(targetValue);

        this.setState({ [targetName]: formattedNumber, cardType: newCardType });
        break;
      case "expiry":
        const formattedExpiry = cardValidator.formatExpiry(targetValue);

        this.setState({ [targetName]: formattedExpiry });
        break;
      default:
        this.setState({
          [targetName]: targetValue
        });
    }
  }

  handleSubmit(event) {
    event.preventDefault();
    console.log(this.state);
    if (!this.isValidForm()) {
      this.displayErrorMessage();
    } else {
      /*
        This is where I would send an API call with the payment information.
        Right now all the info is not encrypted so I would encrypt it before
        sending it to the backend to deal with it.
      */
      alert("You submitted the form correctly");
      this.clearForm();
    }
  }

  displayErrorMessage() {
    const { cvc, name, expiry, number, cardType } = this.state;
    const cardValidator = creditCards.validators[cardType];

    if (!cardValidator.isValidName(name)) {
      this.setState({ errorMessage: "Please input a name" });
    } else if (!cardValidator.isValidNumber(number)) {
      this.setState({ errorMessage: "Invalid card number" });
    } else if (!cardValidator.isValidExpiry(expiry)) {
      this.setState({ errorMessage: "Invalid expiration" });
    } else if (!cardValidator.isValidCVC(cvc)) {
      this.setState({ errorMessage: "Invalid CVC" });
    } else {
      this.setState({
        errorMessage:
          "Invalid credit type, please enter a card which starts with 4, 34, or 37"
      });
    }
  }

  clearForm() {
    const newState = {};
    const inputNames = Object.keys(this.state);
    inputNames.forEach(inputName => {
      newState[inputName] = "";
    });
    this.setState({
      ...newState
    });
  }

  isValidForm() {
    const { cvc, name, expiry, number, cardType } = this.state;
    let cardValidator = creditCards.validators[cardType];

    return (
      cardValidator.isValidCVC(cvc) &&
      cardValidator.isValidNumber(number) &&
      cardValidator.isValidExpiry(expiry) &&
      cardValidator.isValidName(name) &&
      cardValidator.isValidCardType()
    );
  }

  renderInputs() {
    const { name, number, cvc, expiry, cardType, errorMessage } = this.state;
    const numberLength = creditCards.validators[cardType].numberLength;
    const CVCLength = creditCards.validators[cardType].CVCLength;
    const nameMaxLength = creditCards.validators[cardType].nameMaxLength;
    return (
      <>
        <Input
          maxLength={nameMaxLength}
          name="name"
          onChange={this.handleChange}
          placeholder={staticText.namePlaceholder}
          type="text"
          value={name}
          wrapperClassName="m-b-med credit-card-input flex"
          autocomplete="cc-name"
          required
        />
        <div className="m-b-med credit-card-input flex">
          <CardType type={this.state.cardType} wrapperClassName="flex" />
          <Input
            inputMode="numeric"
            maxLength={numberLength}
            name="number"
            onChange={this.handleChange}
            onKeyDown={this.handleSpace}
            placeholder={staticText.numberPlaceholder}
            type="text"
            value={number}
            wrapperClassName="flex"
            autocomplete="cc-number"
          />
        </div>
        <div className="m-b-xl flex">
          <Input
            inputMode="numeric"
            maxLength="7"
            name="expiry"
            onChange={this.handleChange}
            placeholder={staticText.expiryPlaceholder}
            value={expiry}
            wrapperClassName="m-r-med credit-card-input half flex"
            autocomplete="cc-exp"
          />
          <Input
            maxLength={CVCLength}
            name="cvc"
            onChange={this.handleChange}
            placeholder={staticText.CVCPlaceholder}
            type="text"
            value={cvc}
            wrapperClassName="credit-card-input half flex"
            autocomplete="cc-csc"
          />
        </div>
        <Button
          wrapperClassName="flex-horizontal-center"
          baseClassName="credit-card-button"
          text={staticText.button}
          onClick={this.handleSubmit}
        />
        {errorMessage && (
          <div className=" flex-horizontal-center error">
            <div className="flex-vertical-center">
              {/* <IoIosCloseCircle color="#E95757" /> */}
              <Icon color="#E95757" />
              <p className="p-l-xs error-message">{errorMessage}</p>
            </div>
          </div>
        )}
      </>
    );
  }

  render() {
    return (
      <form className="credit-card-form">
        <div className="m-l-xl m-r-xl">{this.renderInputs()}</div>
      </form>
    );
  }
}

export default CreditCardForm;
