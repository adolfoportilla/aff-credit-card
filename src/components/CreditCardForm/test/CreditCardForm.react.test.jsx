import React from "react";
import { shallow } from "enzyme";

import CreditCardForm from "../CreditCardForm";
import CardType from "../components/CardType";
import Button from "../helpers/Button/Button";
import Input from "../helpers/Input/Input";
import creditCards from "../cards";

describe("CreditCardForm", () => {
  describe("handleChange", () => {
    test("finds correct validator", () => {
      const wrapper = shallow(<CreditCardForm />);
      const event = {
        target: {
          name: "number",
          value: "4"
        }
      };

      wrapper
        .find(Input)
        .first()
        .simulate("change", event);

      expect(wrapper.state().cardType).toEqual(creditCards.types.VISA);
    });

    test("correctly updates name", () => {
      const wrapper = shallow(<CreditCardForm />);
      const event = {
        target: {
          name: "name",
          value: "Test"
        }
      };

      wrapper
        .find(Input)
        .first()
        .simulate("change", event);

      expect(wrapper.state().name).toEqual("Test");
    });

    test("correctly updates number and sets cardType", () => {
      const wrapper = shallow(<CreditCardForm />);
      const event = {
        target: {
          name: "number",
          value: "123456"
        }
      };

      wrapper
        .find(Input)
        .at(1)
        .simulate("change", event);

      expect(wrapper.state().number).toEqual("1234 56");
      expect(wrapper.state().cardType).toEqual(creditCards.types.DEFAULT);
    });

    test("correctly updates expiry", () => {
      const wrapper = shallow(<CreditCardForm />);
      const event = {
        target: {
          name: "expiry",
          value: "123"
        }
      };

      wrapper
        .find(Input)
        .at(1)
        .simulate("change", event);

      expect(wrapper.state().expiry).toEqual("12 / 3");
    });

    test("correctly updates cvc", () => {
      const wrapper = shallow(<CreditCardForm />);
      const event = {
        target: {
          name: "cvc",
          value: "123"
        }
      };

      wrapper
        .find(Input)
        .first()
        .simulate("change", event);

      expect(wrapper.state().cvc).toEqual("123");
    });
  });

  describe("handleSubmit", () => {
    const mockEvent = {
      preventDefault: jest.fn()
    };

    test("invalid input renders error", () => {
      const wrapper = shallow(<CreditCardForm />);

      wrapper.setState({
        name: "Random name",
        number: "Card number",
        expiry: "MM / YY",
        cvc: "123"
      });
      wrapper.find(Button).simulate("click", mockEvent);

      expect(wrapper.state("errorMessage")).not.toBe("");
      expect(wrapper.find(".error-message")).toHaveLength(1);
    });

    test("valid input does not render error and clears form", () => {
      const wrapper = shallow(<CreditCardForm />);
      jest.spyOn(window, "alert").mockImplementation(() => {});

      wrapper.setState({
        name: "Random name",
        number: "4141 1234 1234 1234",
        expiry: "12 / 20",
        cvc: "123",
        cardType: "VISA"
      });
      wrapper.find(Button).simulate("click", mockEvent);

      expect(window.alert).toHaveBeenCalled();
      expect(wrapper.state("errorMessage")).toBe("");
      expect(wrapper.find(".error-message")).toHaveLength(0);
      expect(wrapper.state("name")).toBe("");
      expect(wrapper.state("number")).toBe("");
      expect(wrapper.state("expiry")).toBe("");
      expect(wrapper.state("cvc")).toBe("");
      expect(wrapper.state("cardType")).toBe("");
    });
  });

  test("clearForm sets state to empty string", () => {
    const wrapper = shallow(<CreditCardForm />);

    wrapper.setState({ number: "1234", expiry: "12 / 20" });
    wrapper.instance().clearForm();

    expect(wrapper.state("number")).toBe("");
    expect(wrapper.state("expiry")).toBe("");
  });

  test("renders basic layout", () => {
    const wrapper = shallow(<CreditCardForm />);

    expect(wrapper.find("form")).toHaveLength(1);
    expect(wrapper.find(CardType)).toHaveLength(1);
    // Inputs for name, number, expiry, and cvc
    expect(wrapper.find(Input)).toHaveLength(4);
    expect(wrapper.find(Button)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });
});
