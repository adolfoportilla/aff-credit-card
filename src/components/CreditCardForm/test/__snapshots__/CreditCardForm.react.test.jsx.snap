// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardForm renders basic layout 1`] = `ShallowWrapper {}`;
